win32ss/user/ntuser/window.c

```diff
@@ -3352,98 +3352,79 @@ NtUserFindWindowEx(HWND hwndParent,
 }
 
 
-/*
- * @implemented
- */
+/* @implemented */
 PWND FASTCALL UserGetAncestor(PWND Wnd, UINT Type)
 {
-   PWND WndAncestor, Parent;
+    PWND Ret, Parent, pwndNode, pwndMessage;
+    PDESKTOP pDesktop = Wnd->head.rpdesk;
+    PWND pwndDesktop = pDesktop->pDeskInfo->spwnd;
 
-   if (UserHMGetHandle(Wnd) == IntGetDesktopWindow())
-   {
-      return NULL;
-   }
+    if (Wnd == pwndDesktop)
+        return NULL;
 
-   switch (Type)
-   {
-      case GA_PARENT:
-         {
-            WndAncestor = Wnd->spwndParent;
-            break;
-         }
+    pwndMessage = pDesktop->spwndMessage;
+    if (Wnd == pwndMessage)
+        return NULL;
 
-      case GA_ROOT:
-         {
-            WndAncestor = Wnd;
-            Parent = NULL;
+    Parent = Wnd->spwndParent;
+    if (!Parent)
```

Comment:
This is a bit like the `IntIsDesktopWindow(Wnd)` test.

Reply:
Yes, but I want to keep optimized.